### ISSUE TITLE
SameDiff: add 'keep dimensions' option for reductions

### ISSUE THREAD

**Issue by AlexDBlack** [deeplearning4j/nd4j#2545]
https://www.tensorflow.org/api_docs/python/tf/reduce_mean

> keep_dims: If true, retains reduced dimensions with length 1.

So [a,b,c].sum(1) would return [a,1,c] instead of [a,c]

This would be quite useful in many cases (backprop, for example).
I don't believe this is supported at the libnd4j level either:  Mean -> Sum -> BaseAccumulation -> legacy op execution.
https://github.com/deeplearning4j/nd4j/blob/master/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/accum/Mean.java

**Comment by agibsonccc** [comment]
@AlexDBlack https://github.com/deeplearning4j/nd4j/blob/79eb921012122cbe903aecb609cf9e960234aa3b/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/BaseAccumulation.java#L53

**Comment by AlexDBlack** [comment]
Ah, good to know. It may simply be that we need sd.sum(...) etc overloads.
@raver119 do we support 'keepDims' at the libnd4j level?

**Comment by raver119** [comment]
Partially. Atm we support it only for ops imported from TF.

**Comment by raver119** [comment]
So ye, the question is: do we want this option for SameDiff as well? If yes, we can find the way to add it.

**Comment by AlexDBlack** [comment]
Yeah, two reasons we want this:
(a) tensorflow supports it
(b) for some ops, it's be really quite useful to keep the reduced axis as size 1 (otherwise we have to manually reshape so it's broadcastable with the original (pre-reduce) input shape)

**Comment by raver119** [comment]
I think i've added that, but need to check it out first before closing.
